packages/flutter/lib/src/painting/image_stream.dart

```diff
@@ -557,7 +562,7 @@ class MultiFrameImageStreamCompleter extends ImageStreamCompleter {
     }
     final Duration delay = _frameDuration - (timestamp - _shownTimestamp);
     _timer = Timer(delay * timeDilation, () {
```

Comment:
Would it make sense guarding the timer with this check too, so that you don't even register the timer if it's already been registered?

Could you possibly achieve this same effect by just doing something like `_timer ??=`?

Reply:
I don't see how guarding just against double timer registration fixes the described issue (e.g in the addListener,removeListener,addListener scenario).

I'm not sure I see a scenario where we will have double timer registration with the current guard given that `removeListener` cancels the timer after removing all listeners.

Can you describe the scenario you're thinking about in which there will be double timer registration?

Reply:
I may be misreading something here, but it looks like this timer can get created multiple times without getting cancelled, which could be leading to multiple registration.

Reply:
I'm probably missing what you see, can you please describe the sequence of events that leads to multiple timer registration?

Reply:
It looks like the constructor of this class calls `_handleAppFrame` when the codec is ready (https://github.com/flutter/flutter/blob/44c127024d1cddcbf0e63652cbf36ba22d6384f1/packages/flutter/lib/src/painting/image_stream.dart#L515), and when the first listener is added it's also called (https://github.com/flutter/flutter/blob/44c127024d1cddcbf0e63652cbf36ba22d6384f1/packages/flutter/lib/src/painting/image_stream.dart#L616).  It seems entirely possible that the `Image` widget ends up making a new image stream which ends up creating a timer and then adding a listener which creates a second timer before the first has finished.

Reply:
Good catch!
The scenario you describe can actually result in the first frame being skipped(as `_nextFrame` can be set to the second codec frame before the next flutter frame is executed).

Fixed it by only starting to decode after both the codec is ready and a frame listener is added.
In practice I'd imagine that this was very unlikely to happen as the first frame listener was likely to be added synchronously right after creating the stream completer at which point `_codec` is still null (as the async call had no chance to complete) thus the second `_decodeNextFrameAndSchedule` will not be invoked (it's guarded by a check that `_codec` is not null).

Anyway with this fixed I don't see a scenario where the timer is registered multiple times simultaneously.